### PR TITLE
Key sequence validation

### PR DESCRIPTION
**Pull request by spireken**

### CODE REVIEW

File: Applications/Spire/Include/Spire/Spire/CancelKeySequenceValidationModel.hpp [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+namespace Spire {
+
+  //! Represents a model for validating cancel key sequences.
+  class CancelKeySequenceValidationModel : public KeySequenceValidationModel {
+    public:
+
+      //! Constructs a CancelKeySequenceValidationModel.
+      CancelKeySequenceValidationModel();
+
+      virtual ~CancelKeySequenceValidationModel() = default;
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Spire/CancelKeySequenceValidationModel.hpp:14]: Remove this destructor.

File: Applications/Spire/Include/Spire/Spire/CancelKeySequenceValidationModel.hpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+      bool is_valid(const QKeySequence& sequence) const override;
+
+    private:
+      std::vector<ValidKeySequences> m_valid_sequences;
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Spire/CancelKeySequenceValidationModel.hpp:19]: Remove this field, this class should not require any data members.

File: Applications/Spire/Include/Spire/Spire/CancelKeySequenceValidationModel.hpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+#include "Spire/Spire/KeySequenceValidationModel.hpp"
+
+namespace Spire {
+
+  //! Represents a model for validating cancel key sequences.
+  class CancelKeySequenceValidationModel : public KeySequenceValidationModel {
+    public:
+
+      //! Constructs a CancelKeySequenceValidationModel.
+      CancelKeySequenceValidationModel();
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Spire/CancelKeySequenceValidationModel.hpp:12]: Don't see a need for a constructor for this class.

File: Applications/Spire/Include/Spire/Spire/KeySequenceValidationModel.hpp [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#ifndef SPIRE_KEY_SEQUENCE_VALIDATION_MODEL_HPP
+#define SPIRE_KEY_SEQUENCE_VALIDATION_MODEL_HPP
+#include <vector>
+#include <QKeySequence>
+#include <QSet>
+
+namespace Spire {
+
+  //! Represents a model for validating key sequences.
+  class KeySequenceValidationModel {
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Spire/KeySequenceValidationModel.hpp:10]: An abstract base class should have a ```virtual``` destructor.

File: Applications/Spire/Include/Spire/Spire/KeySequenceValidationModel.hpp [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    public:
+
+      //! Represents a collection of valid key sequences, where the nth set in
+      //! the list represents the nth key of a potential valid sequence.
+      using ValidKeySequences = std::vector<QSet<Qt::Key>>;
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Spire/KeySequenceValidationModel.hpp:15]: Remove this ```typedef```.

File: Applications/Spire/Include/Spire/Spire/KeySequenceValidationModel.hpp [hunk 6]
```diff
@@ -0,0 +1,25 @@
+#ifndef SPIRE_KEY_SEQUENCE_VALIDATION_MODEL_HPP
+#define SPIRE_KEY_SEQUENCE_VALIDATION_MODEL_HPP
+#include <vector>
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Spire/KeySequenceValidationModel.hpp:3]: Remove the unused includes.

File: Applications/Spire/Include/Spire/Spire/TaskKeySequenceValidationModel.hpp [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+#include "Spire/Spire/KeySequenceValidationModel.hpp"
+
+namespace Spire {
+
+  //! Represents a model for validing task key sequences.
+  class TaskKeySequenceValidationModel : public KeySequenceValidationModel {
+    public:
+
+      //! Constructs a TaskKeySequenceValidationModel.
+      TaskKeySequenceValidationModel();
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Spire/TaskKeySequenceValidationModel.hpp:12]: Same as above.

File: Applications/Spire/Include/Spire/Spire/TestKeySequenceValidationModel.hpp [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+#define SPIRE_TEST_KEY_SEQUENCE_VALIDATION_MODEL_HPP
+#include "Spire/Spire/KeySequenceValidationModel.hpp"
+
+namespace Spire {
+
+  //! Represents a model for testing a KeySequenceInputWidget.
+  class TestKeySequenceValidationModel : public KeySequenceValidationModel {
+    public:
+
+      virtual ~TestKeySequenceValidationModel() = default;
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Spire/TestKeySequenceValidationModel.hpp:11]: Same as above.

File: Applications/Spire/Source/Common/CancelKeySequenceValidationModel.cpp [hunk 9]
```diff
@@ -0,0 +1,24 @@
+#include "Spire/Spire/CancelKeySequenceValidationModel.hpp"
+
+using namespace Spire;
+
+CancelKeySequenceValidationModel::CancelKeySequenceValidationModel()
+  : m_valid_sequences({}) {}
+
+bool CancelKeySequenceValidationModel::is_valid(
+    const QKeySequence& sequence) const {
+  if(sequence.count() == 0) {
```
Comment (kranar) [inline Applications/Spire/Source/Common/CancelKeySequenceValidationModel.cpp:10]: Reimplement this based on the rules about what a valid cancel key sequence is.

File: Applications/Spire/Source/Common/TaskKeySequenceValidationModel.cpp [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+TaskKeySequenceValidationModel::TaskKeySequenceValidationModel()
+  : m_valid_sequences({ValidKeySequences({FUNCTION_KEYS()}),
+      ValidKeySequences({{Qt::Key_Shift, Qt::Key_Alt, Qt::Key_Control},
+      FUNCTION_KEYS()})}) {}
+
+bool TaskKeySequenceValidationModel::is_valid(
+    const QKeySequence& sequence) const {
+  if(sequence.count() == 0) {
```
Comment (kranar) [inline Applications/Spire/Source/Common/TaskKeySequenceValidationModel.cpp:32]: Same as above, reimplement this based on the rules for a valid task key sequence.

**Review by kranar — CHANGES_REQUESTED** [review]

**Comment by spireken** [comment]
It may be beneficial to add methods to the KeySequenceValidationModel header that return certain sets of keys used by the is_valid method:

``const QSet<Qt::Key>& get_function_keys() const``

``const QSet<Qt::Key>& get_modifier_keys() const``

**Comment by kranar** [comment]
Implement it without those for the time being.

**Pull request merged**